### ISSUE TITLE
Inquiry: Best practices for fine-tuning canary-qwen-2.5b (SALM) for ASR tasks

### ISSUE THREAD

**Issue by msh9184** [NVIDIA-NeMo/NeMo#14438]
Hi, I am trying to fine-tune the canary-qwen-2.5b model (using SALM model in speechlm2 recipe) for ASR tasks. 
So far, I can run the basic scripts successfully, but I would greatly appreciate some guidance or recommendations to improve training, validation, inference, parallel training, and data setup.

Specifically, I would be grateful if you could share any best practices or user guidance regarding:

1. Training setup  
   - Effective batch and bucketing strategies (batch sizes, bucketing configurations)  
   - Recommended optimizers, schedulers, and learning rates  
   - Precision choices (e.g., FP16, BF16) and any tips for stable training  

2. Validation details
   - Best practices for measuring WER during training (datasets, scheduling, tools)  
   - Any validation workflows or evaluation tricks to monitor ASR performance effectively  

3. ASR inference & evaluation
   - Recommended post-processing or normalization methods  
   - Step-by-step inference procedures (e.g., recommended `generate(...)` parameters)  

4. Distributed / parallel training 
   - Guidelines for training with FSDP, DDP, or multi-node setups  
   - Useful config patterns, recommended flags, or known caveats  

5. Dataset loader & prompt configuration for SALM
   - Prompt templates (e.g., “Transcribe the following: <|audioplaceholder|>”) or other LLM prompt tips  
   - Tokenizer setups per LLM variant, if relevant  
   - Preferred dataloader formats (e.g., conversation manifests, JSONL manifests) for ASR fine-tuning with SALM  

I believe such guidance would be highly valuable, especially given the lack of official fine-tuning documentation specifically for canary-qwen-2.5b. Thank you in advance for any suggestions or pointers you can provide!

**Comment by ZhiyuLi-Nvidia** [comment]
@nithinraok could you help take a look?

**Comment by aravindMahadevan** [comment]
Can we get an update or answer to this issue? Seems valuable.

**Comment by msh9184** [comment]
@aravindMahadevan You might want to refer to this repo https://github.com/msh9184/NeMo-speechlm2.git I reproduced. It works fine. I hope this helps.

**Comment by pzelasko** [comment]
I know it's a little late but hopefully still can be useful.

> Effective batch and bucketing strategies (batch sizes, bucketing configurations)

Use 1D bucketing + OOMptimizer https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/asr/datasets.html#pushing-gpu-utilization-to-the-limits-with-bucketing-and-oomptimizer

> Recommended optimizers, schedulers, and learning rates

We used AdamW with cosine LR, IIRC the released ckpt was trained for around 90k steps. You can fetch specific hparams from `config.json` in the HF model repo.

> Precision choices (e.g., FP16, BF16) and any tips for stable training

We used bf16 (no AMP), there were no issues with training stability at any point.

> Best practices for measuring WER during training (datasets, scheduling, tools)

Save checkpoints regularly and evaluate each of them with the eval script on as many datasets as you like. I don't like running generation for LLM based models during training as it takes a lot of time.

> Recommended post-processing or normalization methods

We just used WhisperNormalizer, nothing else.

> Step-by-step inference procedures (e.g., recommended generate(...) parameters)

Regular greedy decoding (no special options) works fine out of the box and was used for the HF leaderboard reporting. Maybe you can get some improvement with beam search but I didn't check.

> Guidelines for training with FSDP, DDP, or multi-node setups

You can train this model with pure DDP on 80GB GPUs, and probably even 40GB GPUs. 
FSDP2 is an option to get a little larger batch sizes (in YAML config, change DDPStrategy to ModelParalleStrategy and set data_parallel_rank), but you have to have fast interconnection (eg NVLink) and absolutely max out the batch sizes, otherwise there's a good chance of communication bottlenecks. 

> Useful config patterns, recommended flags, or known caveats

Nothing comes to mind beyond what's already in YAML.

> Prompt templates (e.g., “Transcribe the following: <|audioplaceholder|>”) or other LLM prompt tips

You can customize the prompt per-dataset to guide the model better or add specific instructions about output formatting using `input_cfg` with multiple groups, e.g.
```
- input_cfg:
  - type: multimodal_conversation
    manifest_filepath: ...
    tarred_audio_filepaths: ...
    tags:
      context: “Transcribe the following in lowercase only: <|audioplaceholder|>”
```

> Tokenizer setups per LLM variant, if relevant

Nothing comes to mind.

> Preferred dataloader formats (e.g., conversation manifests, JSONL manifests) for ASR fine-tuning with SALM

Generally it's best to present data in NeMoMultimodalConversation format (https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/asr/datasets.html#configuring-multimodal-dataloading).
But you can re-use existing NeMo ASR format and Lhotse format data by wrapping it in another group `lhotse_as_conversation`, e.g.:

```yaml
- input_cfg:
  - type: lhotse_as_conversation
    input_cfg:
      - type: nemo_tarred
        manifest_filepath: ...
        tarred_audio_filepaths: ...
        tags:
          context: “Transcribe the following in lowercase only: <|audioplaceholder|>”
```

(I realize this requires better docs, and will likely be refactored and improved later this year)
